lib/active_model/serializer/fragment_cache.rb

```diff
@@ -0,0 +1,99 @@
+module ActiveModel
+  class Serializer
+    class FragmentCache
+
+      attr_reader :serializer
+
+      def initialize(adapter, serializer, options, root)
+        @root       = root
+        @options    = options
+        @adapter    = adapter
+        @serializer = serializer
+      end
+
+      def fetch
+        klass = serializer.class
+        serializers = fragment_serializer(@serializer.object.class.name, klass)
+
+        cached_serializer = serializers[:cached].constantize.new(@serializer.object)
+        non_cached_serializer = serializers[:non_cached].constantize.new(@serializer.object)
+
+        cached_hash = @adapter.class.new(cached_serializer, @options).serializable_hash
+        non_cached_hash = @adapter.class.new(non_cached_serializer, @options).serializable_hash
+
+        if @serializer.root && @adapter.class == ActiveModel::Serializer::Adapter::JsonApi
+          cached_hash_root(cached_hash, non_cached_hash)
```

Comment:
Yeah, you are definitely right, there is no sense to have a method to a specific adapter using ```FragmentCache``` abstract as it is.
This is the only change I haven't worked yet, I'll give it some thought yet. But polymorphism indeed seems a good way to go.